### PR TITLE
Refactor XLA and TPU checks across codebase

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Move XLA utilities together with the TPUAccelerator
Deprecate one more redundant utility function
Fix mocks in tests via fixtures

Follow-up to https://github.com/Lightning-AI/lightning/pull/14514
Part of #11826 
Part of #12786
Closes #14982, #14979

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @justusschock @awaelchli @rohitgr7 @tchaton

### CODE REVIEW

**Comment by awaelchli** [comment]
@carmocca just to be sure, you are not mixing up tpu_available with xla_available? These are two different things. One can have xla available on a machine without tpus.

**Comment by carmocca** [comment]
@awaelchli In this PR, `tpu_available` is deleted in favor of `TPUAccelerator.is_available()`.

In most places, we want to check `XLA_AVAILABLE` as that's what needs to be `True` to successfully import `torch_xla`. `TPUAccelerator.is_available()` includes XLA_AVAILABLE because `torch_xla` is necessary to check whether we have TPU hardware.

I don't think I mixed things up, but there might be mistakes. Tests are not fixed yet.

**Comment by carmocca** [comment]
@awaelchli I'll need you to run MPS tests locally with this branch

**Review by awaelchli — APPROVED** [review]
@carmocca all tests pass running with MPS

File: src/lightning_lite/accelerators/tpu.py [hunk 1]
```diff
@@ -59,6 +70,64 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
...
+    # we would have to use `torch_xla.distributed.xla_dist` for
+    # multiple VMs and TPU_CONFIG won't be available, running
+    # `xm.get_xla_supported_devices("TPU")` won't be possible.
+    return (xm.xrt_world_size() > 1) or bool(xm.get_xla_supported_devices("TPU"))
+
+
+_XLA_AVAILABLE = RequirementCache("torch_xla")
+
+
+def tpu_distributed() -> bool:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning_lite/accelerators/tpu.py:123]: I thought somebody might find it useful. We also use it internally.

File: tests/tests_pytorch/callbacks/test_finetuning_callback.py [hunk 2]
```diff
@@ -370,6 +374,8 @@ def test_callbacks_restore(tmpdir):
         expected["maximize"] = False
     if _TORCH_GREATER_EQUAL_1_12:
         expected["foreach"] = None
+    if _TORCH_GREATER_EQUAL_1_13:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply tests/tests_pytorch/callbacks/test_finetuning_callback.py:377]: Yes

File: tests/tests_lite/test_lite.py [hunk 3]
```diff
@@ -371,6 +371,7 @@ def test_setup_dataloaders_replace_standard_sampler(shuffle, strategy):
         pytest.param("gpu", "mps:0", marks=RunIf(mps=True)),
     ],
 )
+@mock.patch.dict(os.environ, os.environ.copy(), clear=True)
```
[the remark this answers is not in the document]
Comment (carmocca) [reply tests/tests_lite/test_lite.py:374]: Because TPU leaks new environment variables and because TPU requires that we don't clear out previous environment variables

**Review by otaj — APPROVED** [review]

File: tests/tests_lite/test_lite.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply tests/tests_lite/test_lite.py:374]: Hmm, in which case would a tpu test need the environment variable be set from a previous test?

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by kaushikb11 — APPROVED** [review]

File: src/lightning_lite/accelerators/tpu.py [hunk 4]
```diff
@@ -59,6 +70,64 @@ def register_accelerators(cls, accelerator_registry: Dict) -> None:
...
+
+    # For the TPU Pod training process, for example, if we have
+    # TPU v3-32 with 4 VMs, the world size would be 4 and as
+    # we would have to use `torch_xla.distributed.xla_dist` for
+    # multiple VMs and TPU_CONFIG won't be available, running
+    # `xm.get_xla_supported_devices("TPU")` won't be possible.
+    return (xm.xrt_world_size() > 1) or bool(xm.get_xla_supported_devices("TPU"))
+
+
+_XLA_AVAILABLE = RequirementCache("torch_xla")
```
Comment (kaushikb11) [inline src/lightning_lite/accelerators/tpu.py:120]: `torch_xla` could be further extended to accelerators other than TPUs in the future.

But for now, it's alright to have in the `accelerators/tpu.py` file

File: src/pytorch_lightning/CHANGELOG.md [hunk 5]
```diff
@@ -141,7 +141,14 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated the internal `pl.core.mixins.DeviceDtypeModuleMixin` class ([#14511](https://github.com/Lightning-AI/lightning/pull/14511), [#14548](https://github.com/Lightning-AI/lightning/pull/14548))
 
 
-- Deprecated all functions in `pytorch_lightning.utilities.xla_device` in favor of `lightning_lite.utilities.xla_device` ([#14514](https://github.com/Lightning-AI/lightning/pull/14514))
+- Deprecated all functions in `pytorch_lightning.utilities.xla_device` ([#14514](https://github.com/Lightning-AI/lightning/pull/14514), [#14550](https://github.com/Lightning-AI/lightning/pull/14550))
+  * Deprecated the internal `inner_f` function
+  * Deprecated the internal `pl_multi_process` function
+  * Deprecated the internal `XLADeviceUtils.xla_available` staticmethod
+  * Deprecated the `XLADeviceUtils.tpu_device_exists` staticmethod in favor of `pytorch_lightning.accelerators.TPUAccelerator.is_available()`
```
Comment (kaushikb11) [inline src/pytorch_lightning/CHANGELOG.md:148]: thanos snap 😄

File: src/lightning_lite/accelerators/tpu.py (hunk 4 above)
Comment (carmocca) [reply src/lightning_lite/accelerators/tpu.py:120]: Totally. It could be moved into the `XLAStrategy` at some point if we want to.
